lib/Parser/Parser.cpp

```diff
@@ -1122,6 +1129,102 @@ bool Parser::parseLine(std::string &ErrStr) {
                   }
                   if (!consumeToken(ErrStr))
                     return false;
+                } else if (CurTok.str() == "range") {
+                  if (!consumeToken(ErrStr))
+                    return false;
+                  if (CurTok.K != Token::Eq) {
+                    ErrStr = makeErrStr(TP, "expected '=' for range as 'range='");
+                    return false;
+                  }
+
+                  if (!consumeToken(ErrStr))
+                    return false;
+                  if (CurTok.K != Token::OpenBracket) {
+                    ErrStr = makeErrStr(TP, "expected '[' to specify lower bound of range");
+                    return false;
+                  }
+
+                  if (!consumeToken(ErrStr))
+                    return false;
+                  if (CurTok.K == Token::Int) {
+                    ErrStr = makeErrStr(TP, "expected lower bound of range without width");
+                    return false;
+                  }
+                  if (CurTok.K != Token::UntypedInt) {
+                    ErrStr = makeErrStr(TP, "expected lower bound of range");
+                    return false;
+                  }
+                  Lower = CurTok.Val;
+
+                  if (Lower.getBitWidth() != InstWidth) {
```

Comment:
the code in this block seems too complicated for what you are trying to accomplish

Reply:
I suggest writing a little helper function that checks if an APInt fits into a different (smaller) bitwidth and then just calling it to validate both the lower and upper bounds

Reply:
in fact it looks like APInt already has the helper functions you need: isintN() and isSignedIntN()

Reply:
I think the easy thing to do here is to call both of these functions, since textual Souper IR is allowed to specify numbers either signed or unsigned

Reply:
Thanks, this API has saved me from writing that ugly code. I added a check: if (isSignedIntN(Width) || isIntN(Width)) {OK} else {ERROR}. Now, in general, i8 will have numbers between -128 to 255, i.e. -128 to 127 and 0 to 255. (though I am not happy with this kind of check here because signed and unsigned have different range). But, surely if we have 256 for an i8, it will return an error. Since Souper IR cannot explicitly specify signed or unsigned variables, should we make this check complicated to see if one of the values of a range is negative, then it must satisfy isSignedIntN() API?

Reply:
right, if it's negative use is SignedIntN and otherwise use the unsigned check. this won't be very complicated. you should have test cases making sure we reject -129 and 256, but accept -128 and 255.